### PR TITLE
RSDK-6334 - add project and module templates to repository

### PR DESCRIPTION
**Pull request by gvaradarajan**
This is working toward the eventual deletion of https://github.com/viamrobotics/micro-rdk-robot-template and https://github.com/viamrobotics/micro-rdk-module-template by folding the templates into this repository and making version updates easier to manage as a consequence.

The resulting commands once this is merged would be:
`cargo generate --git https://github.com/viamrobotics/micro-rdk template/project` (for the robot template)
and
`cargo generate --git https://github.com/viamrobotics/micro-rdk template/module` (for the module template)

For local testing:
`cargo generate --path <path_to_micro_rdk_directory> template/project` (for the robot template)
and
`cargo generate --path <path_to_micro_rdk_directory> template/module` (for the module template)

Introduced some template changes as well for the end user's convenience. I've expounded on these changes in another branch in this [commit](https://github.com/gvaradarajan/micro-rdk/commit/92b54819453f413f3eef9198544690b9c100f7c4) which provides templates for components. Let me know if I should fold those in, or deal with them separately (or not do them at all)

### CODE REVIEW

**Review by acmorrow — COMMENTED** [review]
I think if we do any future mono-repo-ization tasks like this, the first commit in the PR should be just a copy without modification of the files to be added, and then any changes stacked above as additional commits. As is, it is hard to see the changes made to the templates, and there are some meaningful ones.

File: templates/module/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{% else %}
+{% endif %}
+
+pub fn register_models(registry: &mut ComponentRegistry) -> anyhow::Result<(), RegistryError> {
+    {% if starting_component == "Motor" %}registry.register_motor("my_motor", &My{{starting_component}}::from_config){% elsif starting_component == "Base" %}registry.register_base("my_base", &My{{starting_component}}::from_config){% elsif starting_component == "MovementSensor" %}registry.register_movement_sensor("my_movement_sensor", &My{{starting_component}}::from_config){% elsif starting_component == "PowerSensor" %}registry.register_power_sensor("my_power_sensor", &My{{starting_component}}::from_config){% elsif starting_component == "Sensor" %}registry.register_sensor("my_sensor", &My{{starting_component}}::from_config){% elsif starting_component == "Servo" %}registry.register_servo("my_servo", &My{{starting_component}}::from_config){% elsif starting_component == "GenericComponent" %}registry.register_generic_component("my_generic_component", &My{{starting_component}}::from_config){% elsif starting_component == "Encoder" %}registry.register_encoder("my_encoder", &My{{starting_component}}::from_config){% else %}Ok(()){% endif %}
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply templates/module/src/lib.rs:27]: sadly not, I tried many different ways. The end result always involved a crazy amount of new lines and whitespaces added

**Comment by gvaradarajan** [comment]
> I think if we do any future mono-repo-ization tasks like this, the first commit in the PR should be just a copy without modification of the files to be added, and then any changes stacked above as additional commits. As is, it is hard to see the changes made to the templates, and there are some meaningful ones.

A fair point, will do so in the future. The only real changes are in Cargo.toml for micro-rdk and in templates/module/src/lib.rs and templates/module/cargo-generate.toml (where I added conditional imports based on a new option to select a starting component)

File: templates/module/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (acmorrow) [reply templates/module/src/lib.rs:27]: Unfortunate, but OK.

**Review by acmorrow — APPROVED** [review]
LGTM but please wait for @npmenard's thoughts.

**Comment by npmenard** [comment]
is it ready for review?

**Comment by gvaradarajan** [comment]
> is it ready for review?

yes it's ready for review now, just wanted to make sure the way I was doing this wasn't completely off base

File: templates/module/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "{{project-name}}"
+version = "0.1.0"
+edition = "2021"
+authors = ["{{authors}}"]
+resolver = "2"
+
+[dependencies]
+anyhow = {version = "1", features = ["backtrace"]}
+micro-rdk = {git = "https://github.com/viamrobotics/micro-rdk.git", features = ["{{mcu}}"]}
```
Comment (npmenard) [inline templates/module/Cargo.toml:10]: can you add the version of micro-rdk?

File: templates/module/cargo-generate.toml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+[template]
+cargo_generate_version = ">=0.10.0"
+ignore = [".git","README.md"]
+description = "Micro-RDK Module Template"
+
+[placeholders.mcu]
+type = "string"
+prompt = "MCU"
+choices = ["esp32", "native"]
```
Comment (npmenard) [inline templates/module/cargo-generate.toml:9]: let's remove native here

File: templates/project/cargo-generate.toml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+default = ""
+
+[placeholders.ssid]
+type = "string"
+prompt = "Enter wifi name (only 2.4 gHz network are supported)"
+default = "Viam-2G"
+
+[placeholders.pwd]
+type = "string"
+prompt = "enter the wifi password (leave black if you don't want to store it)"
```
Comment (npmenard) [inline templates/project/cargo-generate.toml:24]: nit: blank

File: templates/project/build.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+            }
+            prost_types::value::Kind::BoolValue(v) => {
+                obj.push_str(&format!("BoolValue({})", v.const_val()));
+            }
+        }
+        format!("Kind::{obj}")
+    }
+}
+
+impl const_gen::CompileConst for ComponentConfig {
```
Comment (npmenard) [inline templates/project/build.rs:363]: this should me removed so let's get rid of it

**Review by npmenard — APPROVED** [review]
LGTM just a couple of nits

**Pull request merged**